### DISCUSSION TITLE
paperwallet not working in real time

### DISCUSSION THREAD

**Discussion by yueismyname** [rodrigo-brito/ninjabot#322] [Q&A]
ninjabot git:(v0.3.6) ✗ go run examples/paperwallet/paperwallet.go
INFO[2024-04-16 15:30] [SETUP] Using Binance exchange               
INFO[2024-04-16 15:30] [SETUP] Using paper wallet                   
INFO[2024-04-16 15:30] [SETUP] Initial Portfolio = 10000.000000 USDT 
Chart available at http://localhost:8080
INFO[2024-04-16 15:30] [SETUP] preloading 22 candles for BNBUSDT-1m 
INFO[2024-04-16 15:30] Bot started.                                 
INFO[2024-04-16 15:30] Connecting to the exchange.                  
INFO[2024-04-16 15:30] Data feed connected.                         
ERRO[2024-04-16 15:30] dataFeedSubscription/start: EOF 

the program just hang there, no more candles coming out,
Do anyone know what's going on?

**Comment by yueismyname** [comment]
Network Issue

**Comment by adamantmm** [comment]
If the bot starts, preloads candles, then sits idle on the websocket with no new candles, it is usually the feed subscription rather than PaperWallet itself.

PaperWallet only reacts to `OnCandle` — it does not pull the book on its own. Things that commonly stall realtime paper mode:

1. Pair/interval mismatch between what you subscribed and what the strategy expects (e.g. `BNBUSDT` vs `BNBUSDT`, or 1m strategy on a 15m feed).
2. Websocket connected but `Candle.Complete` never flips true for that exchange’s kline payload, so the strategy never fires.
3. IP / geo blocks on the Binance public stream — connection succeeds, then no messages.

Quick check: log every candle in the feed callback and confirm `Complete == true` is arriving. If candles flow and orders still do not, then look at the strategy side; if candles stop after preload, it is the websocket path.
